server/src/main/java/org/apache/iotdb/db/mpp/schedule/FragmentInstanceManager.java

```diff
@@ -0,0 +1,286 @@
+  private final IndexedBlockingQueue<FragmentInstanceTask> timeoutQueue;
+  private final Set<FragmentInstanceTask> blockedTasks;
+  private final Map<String, Set<FragmentInstanceTask>> queryMap;
+  private final ITaskScheduler scheduler;
+
+  private static final int MAX_CAPACITY = 1000; // TODO: load from config files
+  private static final int WORKER_THREAD_NUM = 4; // TODO: load from config files
+  private static final int QUERY_TIMEOUT_MS = 10000; // TODO: load from config files or requests
+  private final ThreadGroup workerGroups;
+  private final List<AbstractExecutor> threads;
+
+  public FragmentInstanceManager() {
+    this.readyQueue =
+        new L2PriorityQueue<>(
+            MAX_CAPACITY,
+            new FragmentInstanceTask.SchedulePriorityComparator(),
+            new FragmentInstanceTask());
+    this.timeoutQueue =
+        new L1PriorityQueue<>(
+            MAX_CAPACITY,
+            new FragmentInstanceTask.SchedulePriorityComparator(),
+            new FragmentInstanceTask());
+    this.queryMap = new ConcurrentHashMap<>();
+    this.blockedTasks = Collections.synchronizedSet(new HashSet<>());
+    this.scheduler = new Scheduler();
+    this.workerGroups = new ThreadGroup("ScheduleThreads");
+    this.threads = new ArrayList<>();
+  }
+
+  @Override
+  public void start() throws StartupException {
+    for (int i = 0; i < WORKER_THREAD_NUM; i++) {
+      AbstractExecutor t =
+          new FragmentInstanceTaskExecutor(
+              "Worker-Thread-" + i, workerGroups, readyQueue, scheduler);
+      threads.add(t);
+      t.start();
+    }
+    AbstractExecutor t =
+        new FragmentInstanceTimeoutSentinel(
+            "Sentinel-Thread", workerGroups, timeoutQueue, scheduler);
+    threads.add(t);
+    t.start();
+  }
+
+  @Override
+  public void stop() {
+    this.threads.forEach(
+        t -> {
+          try {
+            t.close();
+          } catch (IOException e) {
+            // Only a field is set, there's no chance to throw an IOException
+          }
+        });
+  }
+
+  @Override
+  public ServiceType getID() {
+    return ServiceType.FRAGMENT_INSTANCE_MANAGER_SERVICE;
+  }
+
+  @Override
+  public void submitFragmentInstances(String queryId, List<ExecFragmentInstance> instances) {
+    Set<FragmentInstanceTask> tasks =
+        instances.stream()
+            .map(
+                v ->
+                    new FragmentInstanceTask(v, QUERY_TIMEOUT_MS, FragmentInstanceTaskStatus.READY))
+            .collect(Collectors.toSet());
+    queryMap.put(queryId, Collections.synchronizedSet(tasks));
+    for (FragmentInstanceTask task : tasks) {
+      task.lock();
+      try {
+        timeoutQueue.push(task);
+        readyQueue.push(task);
+      } finally {
+        task.unlock();
+      }
+    }
+  }
+
+  @Override
+  public void abortQuery(String queryId) {
+    Set<FragmentInstanceTask> queryRelatedTasks = queryMap.remove(queryId);
+    if (queryRelatedTasks != null) {
```

Comment:
If the tasks for a query is being added to the `readyQueue` on line 129 and a abortQuery command for this query is triggered at the same time, the abort command may only clear part of the tasks.

I am not sure whether this scenario will lead to some unexpected results or not...

Reply:
Good catch! I have fixed by re-check the status after getting the lock in `submitFragmentInstance`.